src/qfusion.cpp

```diff
@@ -151,16 +151,38 @@ void QFusion::FlushBit(const bitLenInt& qubitIndex)
     if (bfr) {
         if (bfr->controls.size() == 0) {
             // If this bit has a buffer, with nothing controlling it, we just flush this bit.
-            qReg->ApplySingleBit(bfr->matrix.get(), true, qubitIndex);
+            if (bfr->isArithmetic) {
```

Comment:
It makes me sad, seeing these special cases.  Can that logic be pushed into the BitBuffer as an “Apply”, so that it’s self contained within it’s domain?  Something like:
```
if (bfr) {
    bfr->Apply(qReg)
}
...
BitBuffer::Apply(QInterface qReg)
{
    if (bfr->controls.size() == 0) {
        bfr->ApplyIndependent(qReg);
    } else {
        bfr->ApplyControlled(qReg);
    }
}
```

More functions will improve this code, I think.

Reply:
Done. One "Apply(...)" method.